src/lib/contracts.ts

```diff
@@ -0,0 +1,122 @@
+import { isLeft } from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+import { reporter } from 'io-ts-reporters';
+import * as _ from 'lodash';
+
+import { Blueprint, Contract, ContractObject } from '@balena/contrato';
+
+import constants = require('./constants');
+import { InternalInconsistencyError } from './errors';
+import * as osRelease from './os-release';
+import supervisorVersion = require('./supervisor-version');
+
+export { ContractObject };
+
+export interface ServiceContracts {
+	[serviceName: string]: ContractObject;
+}
+
+export async function containerContractsFulfilled(
+	serviceContracts: ServiceContracts,
+): Promise<{ valid: boolean; unmetServices: string[] }> {
+	const containers = _.values(serviceContracts);
+
+	const osContract = new Contract({
+		slug: 'balenaOS',
+		type: 'sw.os',
+		name: 'balenaOS',
+		version: await osRelease.getOSSemver(constants.hostOSVersionPath),
+	});
+
+	const supervisorContract = new Contract({
+		slug: 'balena-supervisor',
+		type: 'sw.supervisor',
+		name: 'balena-supervisor',
+		version: supervisorVersion,
+	});
+
+	const blueprint = new Blueprint(
+		{
+			'sw.os': 1,
+			'sw.supervisor': 1,
+			'sw.container': '1+',
+		},
+		{
+			type: 'sw.runnable.configuration',
+			slug: '{{children.sw.container.slug}}',
+		},
+	);
+
+	const universe = new Contract({
+		type: 'meta.universe',
+	});
+
+	universe.addChildren(
+		[osContract, supervisorContract].concat(
+			containers.map(c => new Contract(c)),
+		),
+	);
+
+	const solution = blueprint.reproduce(universe);
+
+	if (solution.length > 1) {
+		throw new InternalInconsistencyError(
+			'More than one solution available for container contracts when only one is expected!',
```

Comment:
How could we get multiple OS versions?

Reply:
We can't :)
I'm just theorizing to confirm I correctly understand how this snippet works.

Reply:
Oh I sorry I misunderstood your question. Because of the `1+` in the container membership of the blueprint, we can rely on this. I figured it was worth checking against because I accidentally made a typo in development and this occurred. I'd rather see this in development, rather than debug why suddenly all contracts are failing :P